### ISSUE TITLE
Relative imports failing in dev mode

### ISSUE THREAD

**Issue by dbismut** [preconstruct/preconstruct#378]
Hi again! Following up on #375, same repo but this time I've added another package to the workspace, which is a Vitejs powered app consuming the lib.

Repro repo: https://github.com/dbismut/preconstruct-repro

In a few words, the setup is like so:

```json
"workspaces": [
  "packages/a",
  "b"
],
"preconstruct": {
  "packages": [
    "packages/*"
  ]
},
```

`b` is the Vitejs app, `a` is the lib handled by `preconstruct`. When starting the Vite app, consuming the lib `a` fails with the following error:

```
File: /Users/david/repos/preconstruct-repro/packages/a/dist/myrepo-a.esm.js
  1  |  import { data } from "./data";
     |                        ^
  2  |  
  3  |  export function fn1() {
      at formatError (/Users/david/repos/preconstruct-repro/node_modules/vite/dist/node/chunks/dep-1bdbec90.js:45829:46)
      at TransformContext.error (/Users/david/repos/preconstruct-repro/node_modules/vite/dist/node/chunks/dep-1bdbec90.js:45825:19)
      at normalizeUrl (/Users/david/repos/preconstruct-repro/node_modules/vite/dist/node/chunks/dep-1bdbec90.js:47314:26)
      at async TransformContext.transform (/Users/david/repos/preconstruct-repro/node_modules/vite/dist/node/chunks/dep-1bdbec90.js:47443:57)
      at async Object.transform (/Users/david/repos/preconstruct-repro/node_modules/vite/dist/node/chunks/dep-1bdbec90.js:46027:30)
      at async transformRequest (/Users/david/repos/preconstruct-repro/node_modules/vite/dist/node/chunks/dep-1bdbec90.js:61624:29)
      at async /Users/david/repos/preconstruct-repro/node_modules/vite/dist/node/chunks/dep-1bdbec90.js:61732:32
```

I guess the problem is that relative imports with aliases seem to fail? I'm not expert enough in build tools, but maybe a strategy like the one you're having with commonjs would solve this?

```js
// myrepo-a.esm.js
export * from '../src/index.js'
```

Note that running `preconstruct watch` prior to launching the Vite app solves this.

**Comment by dbismut** [comment]
By the way, with our main repo Leva (https://github.com/pmndrs/leva) which uses Typescript, aliasing `dist/*.esm.js` to `../src/index.ts` generates errors — on top of the relative import not resolving issue I mentioned above — as some of our `index.ts` files have typescript specific syntax such as `export type`, which isn't accepted in pure javascript files.

I've tried manually to replace the alias with a plain esm.js files as I was suggesting above, ie:

```js
export * from '../src/index.ts'
```

This seems to work. I can try to make a PR if you point me to the right direction as I'm not very familiar with build tools internals.

**Comment by Andarist** [comment]
Note that I'm not commenting on the latter part mentioning TS syntax etc as I haven't really digested that yet. It also feels like maybe either some misconfiguration on your side (or a specific issue with Vite perhaps) or like a separate issue.

As to the `"./data"` not being resolved correctly from `"/preconstruct-repro/packages/a/dist/myrepo-a.esm.js"`:

We might notice here that the importer is in the `dist` directory - but it's actually a linked file and thus it should be resolved to the real path (in the `src` directory). This is IMHO a problem with Vite's resolver as this works correctly with most of the popular tools.

This particular issue can be fixed by adjusting the body of this condition:
https://github.com/vitejs/vite/blob/39820b96cec2c672e849e160b92b42c979b85285/packages/vite/src/node/plugins/resolve.ts#L348-L355
to:
```js
resolved = fs.realpathSync(resolved)
return { id: resolved }
```
but I'm not super convinced that this is the appropriate place for the fix - the resolving logic in Vite is quite complex and has some layers. So it would really be best for them to assess at what level this should be fixed. It's also worth noting that it might be worthwhile to use `fs.realpathSync.native` if it's available in the given node version.

**Comment by dbismut** [comment]
@Andarist I think I understand, thanks for digging. Indeed I've added a snowpack app inside my repro repo and it does resolve the relative import (so I guess I should file an issue with Vite, but I think it used to fail with parcel v1).

But Snowpack just fails to resolve any imports in our actual repo, so I need to understand what can be wrong. I'll do my homework and get back to you. You can close the issue since I think it's unrelated to `preconstruct`.

Thanks again.

**Comment by Andarist** [comment]
If you happen to file a new issue please ping me there as I'm not watching all issues in this repo.

**Comment by dbismut** [comment]
Thanks @Andarist much appreciated. I've just updated the repro repo to Typescript and now the Snowpack bundler fails with the following error:

<img width="995" alt="image" src="https://user-images.githubusercontent.com/5003380/109832919-1f03e700-7c41-11eb-83e4-0e2c18dcae7c.png">

Here is the content of `a.ts.proxy.js`:

```js
export default "/_snowpack/pkg/@myrepo/a.ts";
```

This might be because of Snowpack?

**Comment by Andarist** [comment]
I think you forgot to push out the `c` workspace - `yarn snowpack` can't be executed correctly right now.

**Comment by dbismut** [comment]
@Andarist sorry there seems to have been a problem with the git. It's now pushed, sorry again about this.

**Comment by Andarist** [comment]
When resolving the import specifier (`@myrepo/a`) [here](https://github.com/snowpackjs/snowpack/blob/397f06e60c6a2e70657399598ecdb08b29c70aa3/esinstall/src/index.ts#L263-L266) it resolves it to the correct on-disk location but determines that this is an asset. The resolved value is this:
```json
{
  "loc": "/preconstruct-repro/packages/a/src/index.ts",
  "type":"ASSET"
}
```

Assets are just things that should not be processed - like images and others. I believe this is the beginning of the problem in their implementation. However if I change the type to `'BUNDLE'` then I get a very similar problem:
> Module "./data" could not be resolved by Snowpack (Is it installed?).

So investigating just a little bit further I could see that [this line](https://github.com/snowpackjs/snowpack/blob/397f06e60c6a2e70657399598ecdb08b29c70aa3/esinstall/src/index.ts#L269) for this would create such entry in the `imports` if the type would be `'BUNDLE'`:
```json
{
  "imports": {
    "@myrepo/a": "./@myrepo/a.js"
  }
}
```

which looks awfully wrong as this won't resolve to a `.js` file after all.

This is all I could gather for now - I would report this to the Snowpack team. There are a lot of ways in which a non-relative import specified could resolve to some kind of a development file (it being a regular linked module, a preconstruct link or something else).

**Comment by dbismut** [comment]
Thanks a lot for looking into this @Andarist. So if I understand correctly, even when `ts` files is properly resolved, we have the same problem as with Vitejs and relative imports, correct?

Is there any front-end build tool you know that works correctly with `preconstruct` in dev mode with typescript on? Note that Snowpack does work when working with js, and that Vitejs does work when not using aliases...!

**Comment by Andarist** [comment]
> Thanks a lot for looking into this @Andarist. So if I understand correctly, even when ts files would be properly resolved, we have the same problem as with Vitejs and relative imports, correct?

Not quite - they are similar but not the same. Vite doesn't resolve a symlink correctly while Snowpack *copies* files into its cache directory (those copies might be pre-bundled so technically they wouldn't be copies but I'm not sure how their pipeline actually works) and when I've changed the type to `BUNDLE` the copied files still weren't correct. This didn't involve symlinks though.

> Is there any front-end build tool you know that works correctly with preconstruct in dev mode with typescript on? 

I've successfully used it with bare webpack, CRA, Rollup, and more. That this didn't work with Vite or Snowpack was a surprise to me and I hope they will resolve their issues so the `dev` mode will become usable as-is with them.
